Generate enums for known tags, elements, and namespaces, and use them in the HTML parser

heycam opened the pull request:
#### 321c28bff9f2f53472e4c3a833614f2e6595b4e3
<pre>
Generate enums for known tags, elements, and namespaces, and use them in the HTML parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=244020">https://bugs.webkit.org/show_bug.cgi?id=244020</a>
rdar://problem/98767683

Reviewed by Ryosuke Niwa.

The HTML parser spends a significant amount of time comparing tag names
(AtomStrings produced in the AtomHTMLToken constructor) to QualifiedName
objects in the HTMLNames, etc. namespaces. For example,
HTMLTreeBuilder::processStartTagForInBody has around 80 invocations of
operator==(const AtomString&amp;, const QualifiedName&amp;) to determine how to
handle a given start tag. Since the the qualified names we&apos;re comparing
against are known at compile time, we can help the compiler by instead
using an enum to represent all the different, known tag names the parser
wants to handle.

This is a 1% speedup on Speedometer 2.1.

The patch introduces three generated enums, which look like this:

enum class TagName : uint16_t {
    Unknown,

    // HTML
    abbr,
    acronym,
    ...
    // MathML
    annotation,
    annotation_xml,
    ...
    // SVG
    altGlyph,
    altGlyphDef,
    ...
    // Tags in multiple namespaces
    a,
    font,
    ...
    // Foreign namespace tag names requiring adjustment
    Unadjusted_altglyph,
    Unadjusted_altglyphdef,
    ...
};

enum class ElementName : uint16_t {
    Unknown,
    HTML_abbr,
    HTML_acronym,
    ...
    MathML_annotation,
    MathML_annotation_xml,
    ...
    SVG_altGlyph,
    SVG_altGlyphDef,
    ...
    HTML_a,
    HTML_font,
    ...
    SVG_a,
    SVG_font,
    ...
};

enum class Namespace : uint8_t {
    Unknown,
    HTML,
    MathML,
    SVG,
    ...
};

TagName is used to represent a known tag name as produced by the HTML
tokenizer. The tokenizer produces ASCII lowercase names, and so most of
the TagName enum values correspond to the set of known ASCII lowercase
names that the parser needs to care about. Some tag names require
adjustment when parsing foreign content; those start off as values like
TagName::Unadjusted_textpath and are changed later on by the HTML parser
(if used in appropriate foreign content) to a case-adjusted value like
TagName::textPath. TagName values are only used during parsing, and not
stored in the DOM.

ElementName is used to represent a (namespace URI, local name) pair
corresponding to an known element that the parser can produce. This is
unlike TagName, which represents a single string of characters
comprising a tag name. ElementName values are stored on QualifiedName
objects and so can be accessed through an Element to check its known
name. Whereas TagName has a single enum value for tags like &lt;font&gt;,
which are used in both HTML and SVG, ElementName identifies these two
separately, as ElementName::HTML_font and ElementName::SVG_font.

Separating TagName and ElementName helps make the HTML parser clearer in
what is being checked.

Namespace is used to represent a known namespace URI that the parser
cares about. Like ElementName, it&apos;s also stored on QualifiedName.

The new generated header files -- TagName.h, ElementName.h, and
Namespace.h -- add a number of functions dealing with these enums:

- TagName findTagName(Span&lt;UChar&gt;): an extended version of findHTMLTag
  (introduced in bug 238804), which supports all known tag names, moved
  from HTMLNames.h.

- ElementName findElementName(Namespace, const String&amp;): like findTag, but
  for known ElementNames in a specific namespace; used in the
  QualifiedName constructor when an ElementName argument is not provided,
  such as when a QualifiedName is created in a document.createElement()
  call.

- Namespace findNamespace(const AtomString&amp;): converts a known namespace
  URI to a Namespace.

- TagName tagNameForElement(ElementName): converts from an ElementName to
  its equivalent TagName. The TagName and ElementName enum values are
  aligned so that for tags that correspond to a single element (unlike
  &lt;a&gt; and &lt;font&gt;, for example) the underlying enum value is the same.
  This function is needed for cases where the HTML parser must, for
  example, generate an implied tag corresponding to the element on top of
  the HTMLElementStack.

- ElementName elementNameForTag(Namespace, TagName): converts a TagName into
  an ElementName for the given namespace; e.g.

    elementForTag(Namespace::HTML, TagName::font) == ElementName::HTML_font
    elementForTag(Namespace::HTML, TagName::path) == ElementName::Unknown

- QualifiedName&amp; qualifiedNameForElement(ElementName): looks up the
  QualifiedName object corresponding to the given element name.

- TagName adjustSVGTagName(TagName): performs case adjustment for the given
  tag name in an SVG context, e.g. TagName::Unadjusted_textpath becomes
  TagName::textPath.

- AtomString&amp; tagNameAsString(TagName): returns the string
  corresponding to the given tag name; this is a lookup in an array of
  AtomStrings that is initialized at startup.

The AtomHTMLToken constructor now calls findTagName and stores the
resulting TagName value on the object. Only if it&apos;s TagName::Unknown
does it convert the Span&lt;UChar&gt; tag name into an AtomString. This avoids
needless AtomString refcounting for the common case where we have a
token for a tag with a known name.

make_names.pl, which is responsible for generating HTMLNames, SVGNames,
etc., is substantially refactored and now also generates these new
header files. To generate the enums, the script must parse the tag name
.in files for all namespaces. Instead of using a tag name&apos;s identifier
(name with &quot;-&quot;s replaced by &quot;_&quot;) as the key to %allTags, we change to
using a key made up of the namespace and the tag name&apos;s identifier, such
as &quot;HTML:div&quot; and &quot;SVG:a&quot;. This avoids conflicts between e.g. &quot;a&quot; in the
HTML and SVG namespaces. For consistency, %allAttrs is changed in the
same way. Also, %allTags is renamed to %allElements (and similar changes
for various other variables and subs), since it&apos;s storing element
information rather than just tags.

Since %parameters would be inaccurate when parsing multiple tag name .in
files, it is cleared before each file is read, and a copy of the current
%parameters values is stored on each $allElements{...} and $allAttrs{...}
object.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Build changes to generate {TagName,ElementName,Namespace}.{h,cpp}.

* Source/WebCore/dom/Element.h:
(WebCore::Element::elementName const):
(WebCore::Element::nodeNamespace const):

Expose ElementName and Namespace values from an Element&apos;s QualifiedName.

* Source/WebCore/dom/QualifiedName.h:
(WebCore::add):
(WebCore::QualifiedName::QualifiedNameImpl::namespaceMemoryOffset):
(WebCore::QualifiedName::namespaceURI const):
* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::computeHash const):
* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::QNameComponentsTranslator::equal):
(WebCore::QNameComponentsTranslator::translate):

Rename m_namespace to m_namespaceURI.

(WebCore::QualifiedName::elementName const):
(WebCore::QualifiedName::nodeNamespace const):

Accessors for ElementName and Namespace.

* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::QualifiedNameImpl):

Move constructor to cpp file so we can initialize m_namespace and
m_elementName.

(WebCore::makeComponents):

Helper for the two constructors.

(WebCore::QualifiedName::QualifiedName):

New constructor overload that takes the ElementName and Namespace for
the QualifiedName. These are passed in from HTMLNames::init etc., and in
other places in the HTML parser when when known.

(WebCore::QualifiedName::init):
(WebCore::nullQName):

Pass in ElementName and Namespace values for these pre-defined
QualifiedNames.

* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::QualifiedNameCache::getOrCreate):

New getOrCreate overload that initializes the ElementName and Namespace
on a newly created QualifiedNameImpl. The existing getOrCreate now
calls into findNamespace and findElementName to look up the correct
values when they&apos;re not passed in. (We could make these lazily computed,
but it&apos;s more straightforward if we do it eagerly. The use of
QualifiedNameCache for all QualifiedNameImpl creation means we will not
often have to compute these.)

* Source/WebCore/dom/make_names.pl:
(defaultElementPropertyHash):
(defaultAttrPropertyHash):
(defaultParametersHash):

Pre-compute more details for elements, attributes, and the global parameters
hash upfront.

(nameAvoidingCppKeyword):

Appends a &quot;_&quot; to identifiers that are known to clash with C++ keywords.

(makeKey):

Makes a key for %allElements or %allAttrs based on the namespace and local
name of the item.

(elementsHandler):
(attrsHandler):

Use the new key for %allElements and %allAttrs. Stored directly into
%allElements and %allAttrs instead of going via %parsedTags and
%parsedAttrs.

Move the &quot;-x-webkit-&quot; prefix modification to defaultAttrPropertyHash.

(readNames):

Move some namespace-related assertions and defaulting here, now that we
need to do it for each .in file we read.

(readAttrs):
(readElements):

Capture the contents of %parameters for each file in a new hash
%allCppNamespaces. (Use the name &quot;CPP namespaces&quot; because the keys
are &quot;HTML&quot;, &quot;MathML&quot;, etc., corresponding to the C++ namespaces
&quot;HTMLNames&quot;, &quot;MathML&quot;, etc. that the definitions are in. As
opposed to the DOM namespace itself.)

(elementCount):

Returns the number of namespaces a given element local name appears in.

(collectAllElementsAndAttrsPerNamespace):

Fills %allElementsPerNamespace and %allNamespacesPerElementLocalName based on
all the elements read.

(printMacros):

Accept a $valueField argument, which is an entry to look up on the tag
or element. If undef, the hash key itself is printed.

Avoid flattening $namesRef unnecessarily.

(buildConstructorMap):
(printFunctionTable):

Remove unused mapToTagName support. Use new keys to look up %allElements.

(printConstructorSignature):
(printConstructorInterior):
(printConstructors):
(printJSElementIncludes):
(printElementIncludes):
(printConditionalElementIncludes):
(printFactoryHeaderFile):
(usesDefaultJSWrapper):
(printWrapperFunctions):
(printWrapperFactoryCppFile):

Use new keys to look up %allElements.

(printTagNameCases):

Prints the switch statement cases of a new element factory
createKnownElement overload that takes a TagName.

(printHeaderHead):
(printCppHead):

Remove the unused XXX_NAMES_HIDE_GLOBALS macro. Support #includes being
passed in.

(printTypeHelpers):

Avoid flattening $namesRef unnecessarily. Use new keys to look up
%allElements.

(printNamesHeaderFile):

Pass in &quot;identifier&quot; as the field of interest to printMacros. Remove
findHTMLTag declaration.

(byElementNameOrder):

Sort function that places element local names in a unique namespace
first, sorted by C++ namespace name and local name, followed by all
local names in multiple namespaces, sorted by local name. This sort
order is what allows the conversion between TagName and ElementName to
be simple.

(printTagNameHeaderFile):
(printTagNameCppFile):
(printElementNameHeaderFile):
(printElementNameCppFile):
(printNamespaceNameHeaderFile):
(printNamespaceNameCppFile):

Generate the new files.

(findMaxStringLength):

Renamed from findMaxTagLength.

(candidatesWithStringLength):

Simplified and renamed from tagsWithLength.

(generateFindNameForLength):

Renamed from generateFindTagForLength. Takes an array ref $candidates
that holds { string, enumValue } objects rather than just strings. This
allows the string being matched to differ from the name of the enum to
return.

(generateFindBody):

Factored out from printNamesCppFile.

(printNamesCppFile):

Use new keys to look up %allElements and %allAttrs. Compute the set of
strings to pass to GenerateStrings here, instead of doing it just after
parsing the .in files.

(printDefinitions):

Pass in Namespace and ElementName values to QualifiedName constructors.

(printFactoryCppFile):

Print new createKnownElement overload that takes a TagName.

(valueForName): Deleted.
(namesToStrings): Deleted.
(usesDefaultWrapper): Deleted.
(findMaxTagLength): Deleted.
(tagsWithLength): Deleted.
(generateFindTagForLength): Deleted.

* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/svg/svgtags.in:

Canonicalize names.

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::AtomHTMLToken):

Take a TagName when constructing a fake tag.

(WebCore::AtomHTMLToken::name const):

Return the name corresponding to m_tagName, if it&apos;s set, or m_name
otherwise.

(WebCore::AtomHTMLToken::setTagName):
(WebCore::AtomHTMLToken::setName): Deleted.

Changed setName into setTagName, used for tag case adjustments.

(WebCore::AtomHTMLToken::tagName const):

* Source/WebCore/html/parser/HTMLConstructionSite.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::hasImpliedEndTag):
(WebCore::shouldUseLengthLimit):
(WebCore::causesFosterParenting):

Use ElementName instead of matching a QualifiedName.

(WebCore::HTMLConstructionSite::insertCustomElement):
(WebCore::HTMLConstructionSite::insertForeignElement):

Remove arguments, since HTMLStackItem is changed to extract any
name/namespace from the Element it&apos;s given.

(WebCore::HTMLConstructionSite::insertFormattingElement):

Use TagName instead of a tag&apos;s AtomString.

(WebCore::qualifiedNameForTag):
(WebCore::qualifiedNameForHTMLTag):

Look up a known QualifiedName for a given tag name in the HTML or a
specified namespace. Returns a newly created QualifiedName if it&apos;s not a
known tag name.

(WebCore::HTMLConstructionSite::createElement):

Use new qualifiedNameForTag function.

(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface):

Use the new HTMLElementFactory::createKnownElement overload that takes a
TagName. Use the new qualifiedNameForHTMLTag function.

(WebCore::HTMLConstructionSite::createElementFromSavedToken):

Convert the current HTMLStackItem&apos;s ElementName back into a TagName to
construct the tag to process.

(WebCore::HTMLConstructionSite::generateImpliedEndTagsWithExclusion):

New overload of generateImpliedEndTagsWithExclusion that takes an
ElementName.

(WebCore::HTMLConstructionSite::findFosterSite):

Use ElementNames instead of QualifiedNames.

* Source/WebCore/html/parser/HTMLElementStack.h:

Change various functions to take or return ElementNames instead of
AtomStrings or QualifiedNames.

(WebCore::HTMLElementStack::ElementRecord::namespaceURI const):

Removed since it&apos;s unused.

* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLNames::isRootNode):
(WebCore::HTMLNames::isScopeMarker):
(WebCore::HTMLNames::isListItemScopeMarker):
(WebCore::HTMLNames::isTableScopeMarker):
(WebCore::HTMLNames::isTableBodyScopeMarker):
(WebCore::HTMLNames::isTableRowScopeMarker):
(WebCore::HTMLNames::isButtonScopeMarker):
(WebCore::HTMLNames::isSelectScopeMarker):
(WebCore::HTMLElementStack::pop):
(WebCore::HTMLElementStack::popUntil):
(WebCore::HTMLElementStack::popUntilPopped):
(WebCore::HTMLElementStack::isMathMLTextIntegrationPoint):
(WebCore::HTMLElementStack::isHTMLIntegrationPoint):
(WebCore::HTMLElementStack::pushHTMLHtmlElement):
(WebCore::HTMLElementStack::pushHTMLHeadElement):
(WebCore::HTMLElementStack::pushHTMLBodyElement):
(WebCore::HTMLElementStack::push):
(WebCore::HTMLElementStack::insertAbove):
(WebCore::HTMLElementStack::remove):
(WebCore::HTMLElementStack::topmost const):
(WebCore::isMarker):
(WebCore::HTMLElementStack::inScope const):
(WebCore::HTMLElementStack::inListItemScope const):
(WebCore::HTMLElementStack::inTableScope const):
(WebCore::HTMLElementStack::inButtonScope const):
(WebCore::HTMLElementStack::inSelectScope const):
(WebCore::HTMLElementStack::hasTemplateInHTMLScope const):
(WebCore::HTMLElementStack::popCommon):
(WebCore::HTMLElementStack::removeNonTopCommon):

Use ElementName instead of matching a QualifiedName.

* Source/WebCore/html/parser/HTMLFormattingElementList.h:
* Source/WebCore/html/parser/HTMLFormattingElementList.cpp:
(WebCore::HTMLFormattingElementList::closestElementInScopeWithName):
Use ElementName instead of matching a QualifiedName.

(WebCore::itemsHaveMatchingNames):
(WebCore::HTMLFormattingElementList::ensureNoahsArkCondition):

Factor out name matching, and check by comparing the ElementName first.
(Since HTMLFormattingElementList should only contain known elements, we
shouldn&apos;t need to fall back to the local name / namespace URI
comparison.)

* Source/WebCore/html/parser/HTMLStackItem.h:

Remove the m_namespaceURI and m_localName AtomStrings and replace them
with ElementName and Namespace members. Needing the local name string is
rare and we can go out to the Element to get it if needed.

(WebCore::HTMLStackItem::HTMLStackItem):

Remove name and namespace URI arguments and fetch them from the Element
instead.

(WebCore::HTMLStackItem::isElement const):
(WebCore::HTMLStackItem::isDocumentFragment const):

Check m_node&apos;s type directly.

(WebCore::HTMLStackItem::localName const):
(WebCore::HTMLStackItem::namespaceURI const):

Fetch these directly from the Element.

(WebCore::HTMLStackItem::elementName const):
(WebCore::HTMLStackItem::nodeNamespace const):

Accessors for the enums.

(WebCore::HTMLStackItem::matchesHTMLTag const):
(WebCore::isInHTMLNamespace):

Use Namespace instead of an AtomString.

(WebCore::isNumberedHeaderElement):
(WebCore::isSpecialNode):

Use ElementName instead of matching a QualifiedName.

(WebCore::HTMLStackItem::hasTagName const): Deleted.

* Source/WebCore/html/parser/HTMLTreeBuilder.h:
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::isTableBodyContextTag):
(WebCore::isNonAnchorNonNobrFormattingTag):
(WebCore::isNonAnchorFormattingTag):
(WebCore::HTMLConstructionSite::isFormattingTag):
(WebCore::HTMLTreeBuilder::processFakeStartTag):
(WebCore::HTMLTreeBuilder::processFakeEndTag):

Use TagNames instead of AtomStrings.

(WebCore::isNumberedHeaderTag): Deleted.
(WebCore::isCaptionColOrColgroupTag): Deleted.
(WebCore::isTableCellContextTag): Deleted.

Removed since these are expanded out into switch cases at their call
sites.

(WebCore::isTableBodyContextElement):

Use TagName instead of comparing AtomStrings. Now debug only, since it&apos;s
also expanded out into switch cases at its call sites.

(WebCore::HTMLTreeBuilder::HTMLTreeBuilder):

Check the context element&apos;s ElementName instead of matching its tag
QualifiedName.

(WebCore::HTMLTreeBuilder::processFakePEndTagIfPInButtonScope):

Use ElementName instead of an AtomString local name when checking the
state of the stack, and a TagName when constructing the AtomHTMLToken.

(WebCore::HTMLTreeBuilder::processCloseWhenNestedTag):

Use ElementName instead of an AtomString local name.

(WebCore::adjustSVGTagNameCase):

Call the generated adjustSVGTagNameCase instead of building and looking
up a HashMap for the tag name case adjustment.

(WebCore::HTMLTreeBuilder::processStartTagForInBody):

Use TagName instead of comparing AtomStrings and use ElementName instead
of AtomString local names. Note the flow for applet/embed/object/marquee
is slightly changed due to the use of the switch statement now.

(WebCore::HTMLTreeBuilder::didCreateCustomOrFallbackElement):

Use element name argument, which is now unnecessary since HTMLStackItem
will extract the name itself.

(WebCore::HTMLTreeBuilder::processTemplateEndTag):
(WebCore::HTMLTreeBuilder::processEndOfFileForInTemplateContents):
(WebCore::HTMLTreeBuilder::processColgroupEndTagForInColumnGroup):
(WebCore::HTMLTreeBuilder::closeTheCell):
(WebCore::HTMLTreeBuilder::processStartTagForInTable):
(WebCore::HTMLTreeBuilder::processBodyEndTagForInBody):
(WebCore::HTMLTreeBuilder::callTheAdoptionAgency):
(WebCore::HTMLTreeBuilder::resetInsertionModeAppropriately):
(WebCore::HTMLTreeBuilder::processEndTagForInTableBody):
(WebCore::HTMLTreeBuilder::processEndTagForInRow):
(WebCore::HTMLTreeBuilder::processEndTagForInCell):
(WebCore::HTMLTreeBuilder::processEndTagForInBody):
(WebCore::HTMLTreeBuilder::processCaptionEndTagForInCaption):
(WebCore::HTMLTreeBuilder::processTrEndTagForInRow):
(WebCore::HTMLTreeBuilder::processTableEndTagForInTable):
(WebCore::HTMLTreeBuilder::processEndTagForInTable):
(WebCore::HTMLTreeBuilder::processEndTag):
(WebCore::HTMLTreeBuilder::insertPhoneNumberLink):
(WebCore::disallowTelephoneNumberParsing):
(WebCore::HTMLTreeBuilder::processCharacterBuffer):
(WebCore::HTMLTreeBuilder::processEndOfFile):
(WebCore::HTMLTreeBuilder::defaultForBeforeHTML):
(WebCore::HTMLTreeBuilder::defaultForBeforeHead):
(WebCore::HTMLTreeBuilder::defaultForInHead):
(WebCore::HTMLTreeBuilder::defaultForInHeadNoscript):
(WebCore::HTMLTreeBuilder::defaultForAfterHead):
(WebCore::HTMLTreeBuilder::processStartTagForInHead):
(WebCore::HTMLTreeBuilder::shouldProcessTokenInForeignContent):
(WebCore::HTMLTreeBuilder::processTokenInForeignContent):

Use TagNames and ElementNames instead of AtomStrings and QualifiedNames.

(WebCore::HTMLTreeBuilder::processStartTag):

Use TagNames and ElementNames instead of AtomStrings and QualifiedNames.
Call processFakeEndTag in a few places where we were constructing an
AtomHTMLToken locally and calling processEndTag.

(WebCore::HTMLTreeBuilder::processAnyOtherEndTagForInBody):

Split processAnyOtherEndTagForInBody into two cases: one for known
elements (common), and the other for unknown elements (rare) where we
have to delve into the Element on the HTMLStackItem to match its name.

* Source/WebCore/html/parser/TextDocumentParser.cpp:
(WebCore::TextDocumentParser::insertFakeFormattingElements):

Use TagName instead of AtomString for tag name.

* Source/WebCore/page/ProcessWarming.cpp:
(WebCore::ProcessWarming::initializeNames):

Initialize tag name strings.

Canonical link: <a href="https://commits.webkit.org/254007@main">https://commits.webkit.org/254007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2cae14e717611d471a2b52f4b5f5cd7c2865213

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87666 "failed 2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96866 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150628 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30109 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26231 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91617 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93284 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24325 "Build was cancelled") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74420 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24316 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79288 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79477 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67167 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79475 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27805 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13301 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73186 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27770 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14317 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26039 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2819 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29465 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37211 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76017 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29367 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33594 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16857 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ElementInlines.h:
```diff
@@ -65,6 +66,16 @@ inline Element* Node::parentElement() const
+    return is<Element>(*this) ? downcast<Element>(*this).elementName() : ElementName::Unknown;
+}
+
+inline Namespace Node::ns() const
```
Please spell out namespace.

heycam replied:
`namespace` is a C++ keyword. Do you have a specific suggestion for a name?

rniwa replied:
Maybe `elementNamespace`?

heycam replied:
The patch ensures this field is accurate for attributes too, so `nodeNamespace` sounds good.

rniwa commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -72,30 +72,32 @@ static inline void setAttributes(Element& element, AtomHTMLToken& token, ParserC
-        || item.hasTagName(rpTag)
-        || item.hasTagName(rtTag)
-        || item.hasTagName(rtcTag);
+    return item.elementName() == ElementName::HTML_dd
```
Can we use `ElementName::HTML::dd` instead?

heycam replied:
We could do that if the name's slightly different, for example:

```c++
enum class ElementName : uint16_t {
....
    HTML_dd,
    SVG_path,
    ...
};

namespace ElementNames {
namespace HTML {
constexpr ElementName dd = ElementName::HTML_dd;
...
}
namespace SVG {
constexpr ElementName path = ElementName::SVG_path;
}
}
```

then you could write either `ElementName::HTML_dd` or `ElementNames::HTML::dd`. Not sure if that's confusing.

But I'm sympathetic to the goal of reducing visual noise, so I'm happy to entertain suggestions.

heycam replied:
Another option:

```c++
enum class ElementName : uint16_t {
    ...
    HTML_dd,
    SVG_path,
    ...
};

namespace ElementNameHTML {
constexpr ElementName dd = ElementName::HTML_dd;
...
}
namespace ElementNameSVG {
constexpr ElementName path = ElementName::SVG_path;
...
}
}
```

then we can write `ElementNameHTML::dd` in place of `ElementName::HTML_dd`. @rniwa do you like either of these, or have any other suggestion?

heycam replied:
I've done something like this, and I think it reads well. I've left it as a separate commit on top of the main commit, in case I change my mind.

rniwa commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -74,30 +75,50 @@ static inline void setAttributes(Element& element, AtomHTMLToken& token, ParserC
-        || item.hasTagName(rtTag)
-        || item.hasTagName(rtcTag);
+    switch (item.elementName()) {
+    case HTML::dd:
```
This looks nice.

rniwa commented on Source/WebCore/DerivedSources.make:
```diff
@@ -2088,6 +2090,20 @@ $(MATH_ML_GENERATED_PATTERNS) : $(WebCore)/dom/make_names.pl $(WebCore)/bindings
+
+DOM_NAMES_GENERATED_FILES = \
+    DOMNames.cpp \
+    DOMNames.h \
```
Do we really need a single file that contains all attribute names & element names?

heycam replied:
As discussed, I'll split the generated files into separate TagName.h/cpp, ElementName.h/cpp, and Namespace.h/cpp files.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +822,378 @@ END
+    print F "class QualifiedName;\n";
+    print F "\n";
+    print F "enum class TagName : uint16_t {\n";
+    print F "    Unknown = 1,\n";
```
Use 0 for Unknown as we discussed.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +822,378 @@ END
                 my $bufferStart = $currentIndex > 0 ? "buffer.data() + $currentIndex" : "buffer.data()";
-                print F "${indent}static constexpr UChar ${tag}Rest[] = { ";
+                print F "${indent}static constexpr characterType rest[] = { ";
                 for (my $index = $currentIndex; $index < $length; $index = $index + 1) {
```
Can we use "name" string instead of 'n', 'a', 'm', 'e' for better readability?

heycam replied:
We can. (This big switch statement generation is existing code, so I might address that as a followup.)

heycam replied:
Mentioned this in https://bugs.webkit.org/show_bug.cgi?id=244524.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +822,378 @@ END
                 }
                 print F "};\n";
-                print F "${indent}if (!memcmp($bufferStart, ${tag}Rest, $lengthToCompare * sizeof(UChar))) {\n";
+                print F "${indent}if (!memcmp($bufferStart, rest, $lengthToCompare * sizeof(characterType))) {\n";
```
We shouldn't have curly braces around a single line statement.

heycam replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=244524 for this.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +822,378 @@ END
             if ($lengthToCompare == 1) {
-                my $letter = substr($tag, $currentIndex, 1);
+                my $letter = substr($string, $currentIndex, 1);
                 print F "${indent}if (buffer[$currentIndex] == '$letter') {\n";
```
We shouldn't have curly braces around a single line statement.

rniwa replied:
Doesn't seem like this has been addressed?

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -915,26 +1211,158 @@ sub printNamesCppFile
+    print F "        return;\n";
+    print F "\n";
+    print F "    s_tagNames.construct();\n";
+    print F "    AtomString* tagNamesEntry = s_tagNames->begin() + 2;\n";
```
Where does this magic 2 come from?

heycam replied:
It was due to skipping past element 0 (not used) and 1 (for `Unknown`). Per your previous comment, I'll make the `Unknown` values 0, which means this 2 will become a 1. And I'll add a comment in the generated code to indicate where that comes from.

rniwa commented on Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
```diff
@@ -290,7 +283,7 @@ HTMLTreeBuilder::HTMLTreeBuilder(HTMLDocumentParser& parser, DocumentFragment& f
     m_tree.openElements().pushRootNode(HTMLStackItem(fragment));
 
-    if (contextElement.hasTagName(templateTag))
+    if (contextElement.elementName() == HTML::template_)
```
This is sad but there is no way around it.

rniwa commented on Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
```diff
@@ -444,22 +437,23 @@ void HTMLTreeBuilder::processFakeCharacters(const String& characters)
 bool isDdOrDt(const HTMLStackItem& item)
 {
-    return item.hasTagName(ddTag) || item.hasTagName(dtTag);
+    return item.elementName() == HTML::dd
```
I think putting them on a single line would read better when there are only a couple of items.

rniwa commented on Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
```diff
@@ -1336,40 +1375,50 @@ void HTMLTreeBuilder::processStartTag(AtomHTMLToken&& token)
+    case InsertionMode::TemplateContents: {
+        InsertionMode insertionMode = InsertionMode::TemplateContents;
+        switch (token.tagName()) {
+        case TagName::base:
```
Where did base, basefont, bgsound, etc... come from?

heycam replied:
Per discussion, this makes us match the HTML spec more closely (though it's not a behavior change). I'll split this out; filed https://bugs.webkit.org/show_bug.cgi?id=244525 for this.

rniwa commented on Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
```diff
@@ -1414,9 +1464,42 @@ void HTMLTreeBuilder::processAnyOtherEndTagForInBody(AtomHTMLToken&& token)
     }
 }
 
+void HTMLTreeBuilder::processAnyOtherEndTagForUnknownElementInBody(AtomHTMLToken&& token)
```
Can we use template & some lambda to share the code between the two?

heycam replied:
Yes we can.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -247,70 +301,56 @@ sub defaultInterfaceName
-        $value = $name;
-        $value =~ s/_/-/g;
-    }
+    my %cppKeywords = map { $_ => 1 } qw(char class default for operator switch template);
```
C++ has a lot more keywords than this right? Is this meant to be just those we care about?
Probably a good idea to add a comment explaining why this subset is picked.

heycam replied:
Yes I just chose the ones that actually conflict with our current set of names. Actually some of those are attributes, and don't need to be there right now. I'll remove those extra ones and add a comment here.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -247,70 +301,56 @@ sub defaultInterfaceName
     my ($tag, $property, $value) = @_;
 
-    $tag =~ s/-/_/g;
+    my $tagKey = makeKey($parameters{namespace}, $tag);
```
Isn't this more of an element key since we're talking about namespace + tag?

heycam replied:
Indeed, you are right. I think that implies we should rename %allTags to %allElements, and various other similar names.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +833,562 @@ END
+    print F "\n";
+    print F "// Windows headers define this macro, which conflicts with an HTML tag name.\n";
+    print F "#if defined(small)\n";
+    print F "#undef small\n";
```
Why not do the same thing as other keywords and just append _ instead?
undef'ing windows macro sounds like a trouble down the line
when someone inevitably include these headers in Windows port.

heycam replied:
Sure, can do that.

rniwa commented on Source/WebCore/DerivedSources.make:
```diff
@@ -2084,7 +2090,68 @@ MATH_ML_GENERATED_PATTERNS = $(subst .,%,$(MATH_ML_GENERATED_FILES))
+
+# --------
+
+# DOM name enums
```
Maybe update this comment now that we don't have DOMNames.h/cpp?

rniwa commented on Source/WebCore/bindings/scripts/StaticString.pm:
```diff
@@ -29,7 +29,6 @@ use Hasher;
 sub GenerateStrings($)
 {
     my $stringsRef = shift;
-    my %strings = %$stringsRef;
```
This change seems unrelated. If it's a cleanup, maybe we want to do that in a separate followup.

heycam replied:
Will split out.

rniwa commented on Source/WebCore/bindings/scripts/StaticString.pm:
```diff
@@ -60,13 +58,12 @@ END
 sub GenerateStringAsserts($)
 {
     my $stringsRef = shift;
-    my %strings = %$stringsRef;
```
Ditto.

rniwa commented on Source/WebCore/dom/QualifiedName.h:
```diff
@@ -45,6 +48,7 @@ class QualifiedName {
     WTF_MAKE_STRUCT_FAST_ALLOCATED_WITH_HEAP_IDENTIFIER(QualifiedName);
 public:
     class QualifiedNameImpl : public RefCounted<QualifiedNameImpl> {
+        friend class QualifiedName;
```
We usually declare friend classes at the end instead of the beginning.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -167,75 +169,126 @@ END
+        identifier => $identifier,
+        tagEnumValue => $tagEnumValue,
+        elementEnumValue => "$parameters{namespace}_$identifier",
+        unadjustedTagEnumValue => $requiresAdjustment ? "Unadjusted_" . lc($identifier) : "",
```
As we discussed before, we should probably use a more descriptive prefix like UnadjectedCase_.
I think suffix like altglyphCaseUnadjusted would read best.

heycam replied:
Ah yes, I forgot.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +833,552 @@ END
+        my $tagEnumValue = $allElements{$elementKey}{tagEnumValue};
+
+        my $sectionComment =
+            elementCount($localName) > 1 && !$seenMultiNamespaceTags ? "\n    // Tags in multiple namespaces\n" :
```
":" should be on the beginning of the next line.

rniwa commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -759,145 +833,552 @@ END
+        print F "TagName adjust${namespace}TagName(TagName);\n";
+    }
+    print F "\n";
+    print F "inline const AtomString& getTagNameAsString(TagName tagName)\n";
```
This should be tagNameAsString or stringForTagName.
"get" prefix should be only used when there is an out argument:
https://webkit.org/code-style-guidelines/#names-setter-getter

rniwa reviewed (approved):
r=me

heycam committed:
Generate enums for known tags, elements, and namespaces, and use them in the HTML parser
https://bugs.webkit.org/show_bug.cgi?id=244020
rdar://problem/98767683

Reviewed by Ryosuke Niwa.

The HTML parser spends a significant amount of time comparing tag names
(AtomStrings produced in the AtomHTMLToken constructor) to QualifiedName
objects in the HTMLNames, etc. namespaces. For example,
HTMLTreeBuilder::processStartTagForInBody has around 80 invocations of
operator==(const AtomString&, const QualifiedName&) to determine how to
handle a given start tag. Since the the qualified names we're comparing
against are known at compile time, we can help the compiler by instead
using an enum to represent all the different, known tag names the parser
wants to handle.

This is a 1% speedup on Speedometer 2.1.

The patch introduces three generated enums, which look like this:

enum class TagName : uint16_t {
    Unknown,

    // HTML
    abbr,
    acronym,
    ...
    // MathML
    annotation,
    annotation_xml,
    ...
    // SVG
    altGlyph,
    altGlyphDef,
    ...
    // Tags in multiple namespaces
    a,
    font,
    ...
    // Foreign namespace tag names requiring adjustment
    Unadjusted_altglyph,
    Unadjusted_altglyphdef,
    ...
};

enum class ElementName : uint16_t {
    Unknown,
    HTML_abbr,
    HTML_acronym,
    ...
    MathML_annotation,
    MathML_annotation_xml,
    ...
    SVG_altGlyph,
    SVG_altGlyphDef,
    ...
    HTML_a,
    HTML_font,
    ...
    SVG_a,
    SVG_font,
    ...
};

enum class Namespace : uint8_t {
    Unknown,
    HTML,
    MathML,
    SVG,
    ...
};

TagName is used to represent a known tag name as produced by the HTML
tokenizer. The tokenizer produces ASCII lowercase names, and so most of
the TagName enum values correspond to the set of known ASCII lowercase
names that the parser needs to care about. Some tag names require
adjustment when parsing foreign content; those start off as values like
TagName::Unadjusted_textpath and are changed later on by the HTML parser
(if used in appropriate foreign content) to a case-adjusted value like
TagName::textPath. TagName values are only used during parsing, and not
stored in the DOM.

ElementName is used to represent a (namespace URI, local name) pair
corresponding to an known element that the parser can produce. This is
unlike TagName, which represents a single string of characters
comprising a tag name. ElementName values are stored on QualifiedName
objects and so can be accessed through an Element to check its known
name. Whereas TagName has a single enum value for tags like <font>,
which are used in both HTML and SVG, ElementName identifies these two
separately, as ElementName::HTML_font and ElementName::SVG_font.

Separating TagName and ElementName helps make the HTML parser clearer in
what is being checked.

Namespace is used to represent a known namespace URI that the parser
cares about. Like ElementName, it's also stored on QualifiedName.

The new generated header files -- TagName.h, ElementName.h, and
Namespace.h -- add a number of functions dealing with these enums:

- TagName findTagName(Span<UChar>): an extended version of findHTMLTag
  (introduced in bug 238804), which supports all known tag names, moved
  from HTMLNames.h.

- ElementName findElementName(Namespace, const String&): like findTag, but
  for known ElementNames in a specific namespace; used in the
  QualifiedName constructor when an ElementName argument is not provided,
  such as when a QualifiedName is created in a document.createElement()
  call.

- Namespace findNamespace(const AtomString&): converts a known namespace
  URI to a Namespace.

- TagName tagNameForElement(ElementName): converts from an ElementName to
  its equivalent TagName. The TagName and ElementName enum values are
  aligned so that for tags that correspond to a single element (unlike
  <a> and <font>, for example) the underlying enum value is the same.
  This function is needed for cases where the HTML parser must, for
  example, generate an implied tag corresponding to the element on top of
  the HTMLElementStack.

- ElementName elementNameForTag(Namespace, TagName): converts a TagName into
  an ElementName for the given namespace; e.g.

    elementForTag(Namespace::HTML, TagName::font) == ElementName::HTML_font
    elementForTag(Namespace::HTML, TagName::path) == ElementName::Unknown

- QualifiedName& qualifiedNameForElement(ElementName): looks up the
  QualifiedName object corresponding to the given element name.

- TagName adjustSVGTagName(TagName): performs case adjustment for the given
  tag name in an SVG context, e.g. TagName::Unadjusted_textpath becomes
  TagName::textPath.

- AtomString& tagNameAsString(TagName): returns the string
  corresponding to the given tag name; this is a lookup in an array of
  AtomStrings that is initialized at startup.

The AtomHTMLToken constructor now calls findTagName and stores the
resulting TagName value on the object. Only if it's TagName::Unknown
does it convert the Span<UChar> tag name into an AtomString. This avoids
needless AtomString refcounting for the common case where we have a
token for a tag with a known name.

make_names.pl, which is responsible for generating HTMLNames, SVGNames,
etc., is substantially refactored and now also generates these new
header files. To generate the enums, the script must parse the tag name
.in files for all namespaces. Instead of using a tag name's identifier
(name with "-"s replaced by "_") as the key to %allTags, we change to
using a key made up of the namespace and the tag name's identifier, such
as "HTML:div" and "SVG:a". This avoids conflicts between e.g. "a" in the
HTML and SVG namespaces. For consistency, %allAttrs is changed in the
same way. Also, %allTags is renamed to %allElements (and similar changes
for various other variables and subs), since it's storing element
information rather than just tags.

Since %parameters would be inaccurate when parsing multiple tag name .in
files, it is cleared before each file is read, and a copy of the current
%parameters values is stored on each $allElements{...} and $allAttrs{...}
object.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Build changes to generate {TagName,ElementName,Namespace}.{h,cpp}.

* Source/WebCore/dom/Element.h:
(WebCore::Element::elementName const):
(WebCore::Element::nodeNamespace const):

Expose ElementName and Namespace values from an Element's QualifiedName.

* Source/WebCore/dom/QualifiedName.h:
(WebCore::add):
(WebCore::QualifiedName::QualifiedNameImpl::namespaceMemoryOffset):
(WebCore::QualifiedName::namespaceURI const):
* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::computeHash const):
* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::QNameComponentsTranslator::equal):
(WebCore::QNameComponentsTranslator::translate):

Rename m_namespace to m_namespaceURI.

(WebCore::QualifiedName::elementName const):
(WebCore::QualifiedName::nodeNamespace const):

Accessors for ElementName and Namespace.

* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::QualifiedNameImpl):

Move constructor to cpp file so we can initialize m_namespace and
m_elementName.

(WebCore::makeComponents):

Helper for the two constructors.

(WebCore::QualifiedName::QualifiedName):

New constructor overload that takes the ElementName and Namespace for
the QualifiedName. These are passed in from HTMLNames::init etc., and in
other places in the HTML parser when when known.

(WebCore::QualifiedName::init):
(WebCore::nullQName):

Pass in ElementName and Namespace values for these pre-defined
QualifiedNames.

* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::QualifiedNameCache::getOrCreate):

New getOrCreate overload that initializes the ElementName and Namespace
on a newly created QualifiedNameImpl. The existing getOrCreate now
calls into findNamespace and findElementName to look up the correct
values when they're not passed in. (We could make these lazily computed,
but it's more straightforward if we do it eagerly. The use of
QualifiedNameCache for all QualifiedNameImpl creation means we will not
often have to compute these.)

* Source/WebCore/dom/make_names.pl:
(defaultElementPropertyHash):
(defaultAttrPropertyHash):
(defaultParametersHash):

Pre-compute more details for elements, attributes, and the global parameters
hash upfront.

(nameAvoidingCppKeyword):

Appends a "_" to identifiers that are known to clash with C++ keywords.

(makeKey):

Makes a key for %allElements or %allAttrs based on the namespace and local
name of the item.

(elementsHandler):
(attrsHandler):

Use the new key for %allElements and %allAttrs. Stored directly into
%allElements and %allAttrs instead of going via %parsedTags and
%parsedAttrs.

Move the "-x-webkit-" prefix modification to defaultAttrPropertyHash.

(readNames):

Move some namespace-related assertions and defaulting here, now that we
need to do it for each .in file we read.

(readAttrs):
(readElements):

Capture the contents of %parameters for each file in a new hash
%allCppNamespaces. (Use the name "CPP namespaces" because the keys
are "HTML", "MathML", etc., corresponding to the C++ namespaces
"HTMLNames", "MathML", etc. that the definitions are in. As
opposed to the DOM namespace itself.)

(elementCount):

Returns the number of namespaces a given element local name appears in.

(collectAllElementsAndAttrsPerNamespace):

Fills %allElementsPerNamespace and %allNamespacesPerElementLocalName based on
all the elements read.

(printMacros):

Accept a $valueField argument, which is an entry to look up on the tag
or element. If undef, the hash key itself is printed.

Avoid flattening $namesRef unnecessarily.

(buildConstructorMap):
(printFunctionTable):

Remove unused mapToTagName support. Use new keys to look up %allElements.

(printConstructorSignature):
(printConstructorInterior):
(printConstructors):
(printJSElementIncludes):
(printElementIncludes):
(printConditionalElementIncludes):
(printFactoryHeaderFile):
(usesDefaultJSWrapper):
(printWrapperFunctions):
(printWrapperFactoryCppFile):

Use new keys to look up %allElements.

(printTagNameCases):

Prints the switch statement cases of a new element factory
createKnownElement overload that takes a TagName.

(printHeaderHead):
(printCppHead):

Remove the unused XXX_NAMES_HIDE_GLOBALS macro. Support #includes being
passed in.

(printTypeHelpers):

Avoid flattening $namesRef unnecessarily. Use new keys to look up
%allElements.

(printNamesHeaderFile):

Pass in "identifier" as the field of interest to printMacros. Remove
findHTMLTag declaration.

(byElementNameOrder):

Sort function that places element local names in a unique namespace
first, sorted by C++ namespace name and local name, followed by all
local names in multiple namespaces, sorted by local name. This sort
order is what allows the conversion between TagName and ElementName to
be simple.

(printTagNameHeaderFile):
(printTagNameCppFile):
(printElementNameHeaderFile):
(printElementNameCppFile):
(printNamespaceNameHeaderFile):
(printNamespaceNameCppFile):

Generate the new files.

(findMaxStringLength):

Renamed from findMaxTagLength.

(candidatesWithStringLength):

Simplified and renamed from tagsWithLength.

(generateFindNameForLength):

Renamed from generateFindTagForLength. Takes an array ref $candidates
that holds { string, enumValue } objects rather than just strings. This
allows the string being matched to differ from the name of the enum to
return.

(generateFindBody):

Factored out from printNamesCppFile.

(printNamesCppFile):

Use new keys to look up %allElements and %allAttrs. Compute the set of
strings to pass to GenerateStrings here, instead of doing it just after
parsing the .in files.

(printDefinitions):

Pass in Namespace and ElementName values to QualifiedName constructors.

(printFactoryCppFile):

Print new createKnownElement overload that takes a TagName.

(valueForName): Deleted.
(namesToStrings): Deleted.
(usesDefaultWrapper): Deleted.
(findMaxTagLength): Deleted.
(tagsWithLength): Deleted.
(generateFindTagForLength): Deleted.

* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/svg/svgtags.in:

Canonicalize names.

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::AtomHTMLToken):

Take a TagName when constructing a fake tag.

(WebCore::AtomHTMLToken::name const):

Return the name corresponding to m_tagName, if it's set, or m_name
otherwise.

(WebCore::AtomHTMLToken::setTagName):
(WebCore::AtomHTMLToken::setName): Deleted.

Changed setName into setTagName, used for tag case adjustments.

(WebCore::AtomHTMLToken::tagName const):

* Source/WebCore/html/parser/HTMLConstructionSite.h:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::hasImpliedEndTag):
(WebCore::shouldUseLengthLimit):
(WebCore::causesFosterParenting):

Use ElementName instead of matching a QualifiedName.

(WebCore::HTMLConstructionSite::insertCustomElement):
(WebCore::HTMLConstructionSite::insertForeignElement):

Remove arguments, since HTMLStackItem is changed to extract any
name/namespace from the Element it's given.

(WebCore::HTMLConstructionSite::insertFormattingElement):

Use TagName instead of a tag's AtomString.

(WebCore::qualifiedNameForTag):
(WebCore::qualifiedNameForHTMLTag):

Look up a known QualifiedName for a given tag name in the HTML or a
specified namespace. Returns a newly created QualifiedName if it's not a
known tag name.

(WebCore::HTMLConstructionSite::createElement):

Use new qualifiedNameForTag function.

(WebCore::HTMLConstructionSite::createHTMLElementOrFindCustomElementInterface):

Use the new HTMLElementFactory::createKnownElement overload that takes a
TagName. Use the new qualifiedNameForHTMLTag function.

(WebCore::HTMLConstructionSite::createElementFromSavedToken):

Convert the current HTMLStackItem's ElementName back into a TagName to
construct the tag to process.

(WebCore::HTMLConstructionSite::generateImpliedEndTagsWithExclusion):

New overload of generateImpliedEndTagsWithExclusion that takes an
ElementName.

(WebCore::HTMLConstructionSite::findFosterSite):

Use ElementNames instead of QualifiedNames.

* Source/WebCore/html/parser/HTMLElementStack.h:

Change various functions to take or return ElementNames instead of
AtomStrings or QualifiedNames.

(WebCore::HTMLElementStack::ElementRecord::namespaceURI const):

Removed since it's unused.

* Source/WebCore/html/parser/HTMLElementStack.cpp:
(WebCore::HTMLNames::isRootNode):
(WebCore::HTMLNames::isScopeMarker):
(WebCore::HTMLNames::isListItemScopeMarker):
(WebCore::HTMLNames::isTableScopeMarker):
(WebCore::HTMLNames::isTableBodyScopeMarker):
(WebCore::HTMLNames::isTableRowScopeMarker):
(WebCore::HTMLNames::isButtonScopeMarker):
(WebCore::HTMLNames::isSelectScopeMarker):
(WebCore::HTMLElementStack::pop):
(WebCore::HTMLElementStack::popUntil):
(WebCore::HTMLElementStack::popUntilPopped):
(WebCore::HTMLElementStack::isMathMLTextIntegrationPoint):
(WebCore::HTMLElementStack::isHTMLIntegrationPoint):
(WebCore::HTMLElementStack::pushHTMLHtmlElement):
(WebCore::HTMLElementStack::pushHTMLHeadElement):
(WebCore::HTMLElementStack::pushHTMLBodyElement):
(WebCore::HTMLElementStack::push):
(WebCore::HTMLElementStack::insertAbove):
(WebCore::HTMLElementStack::remove):
(WebCore::HTMLElementStack::topmost const):
(WebCore::isMarker):
(WebCore::HTMLElementStack::inScope const):
(WebCore::HTMLElementStack::inListItemScope const):
(WebCore::HTMLElementStack::inTableScope const):
(WebCore::HTMLElementStack::inButtonScope const):
(WebCore::HTMLElementStack::inSelectScope const):
(WebCore::HTMLElementStack::hasTemplateInHTMLScope const):
(WebCore::HTMLElementStack::popCommon):
(WebCore::HTMLElementStack::removeNonTopCommon):

Use ElementName instead of matching a QualifiedName.

* Source/WebCore/html/parser/HTMLFormattingElementList.h:
* Source/WebCore/html/parser/HTMLFormattingElementList.cpp:
(WebCore::HTMLFormattingElementList::closestElementInScopeWithName):
Use ElementName instead of matching a QualifiedName.

(WebCore::itemsHaveMatchingNames):
(WebCore::HTMLFormattingElementList::ensureNoahsArkCondition):

Factor out name matching, and check by comparing the ElementName first.
(Since HTMLFormattingElementList should only contain known elements, we
shouldn't need to fall back to the local name / namespace URI
comparison.)

* Source/WebCore/html/parser/HTMLStackItem.h:

Remove the m_namespaceURI and m_localName AtomStrings and replace them
with ElementName and Namespace members. Needing the local name string is
rare and we can go out to the Element to get it if needed.

(WebCore::HTMLStackItem::HTMLStackItem):

Remove name and namespace URI arguments and fetch them from the Element
instead.

(WebCore::HTMLStackItem::isElement const):
(WebCore::HTMLStackItem::isDocumentFragment const):

Check m_node's type directly.

(WebCore::HTMLStackItem::localName const):
(WebCore::HTMLStackItem::namespaceURI const):

Fetch these directly from the Element.

(WebCore::HTMLStackItem::elementName const):
(WebCore::HTMLStackItem::nodeNamespace const):

Accessors for the enums.

(WebCore::HTMLStackItem::matchesHTMLTag const):
(WebCore::isInHTMLNamespace):

Use Namespace instead of an AtomString.

(WebCore::isNumberedHeaderElement):
(WebCore::isSpecialNode):

Use ElementName instead of matching a QualifiedName.

(WebCore::HTMLStackItem::hasTagName const): Deleted.

* Source/WebCore/html/parser/HTMLTreeBuilder.h:
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::isTableBodyContextTag):
(WebCore::isNonAnchorNonNobrFormattingTag):
(WebCore::isNonAnchorFormattingTag):
(WebCore::HTMLConstructionSite::isFormattingTag):
(WebCore::HTMLTreeBuilder::processFakeStartTag):
(WebCore::HTMLTreeBuilder::processFakeEndTag):

Use TagNames instead of AtomStrings.

(WebCore::isNumberedHeaderTag): Deleted.
(WebCore::isCaptionColOrColgroupTag): Deleted.
(WebCore::isTableCellContextTag): Deleted.

Removed since these are expanded out into switch cases at their call
sites.

(WebCore::isTableBodyContextElement):

Use TagName instead of comparing AtomStrings. Now debug only, since it's
also expanded out into switch cases at its call sites.

(WebCore::HTMLTreeBuilder::HTMLTreeBuilder):

Check the context element's ElementName instead of matching its tag
QualifiedName.

(WebCore::HTMLTreeBuilder::processFakePEndTagIfPInButtonScope):

Use ElementName instead of an AtomString local name when checking the
state of the stack, and a TagName when constructing the AtomHTMLToken.

(WebCore::HTMLTreeBuilder::processCloseWhenNestedTag):

Use ElementName instead of an AtomString local name.

(WebCore::adjustSVGTagNameCase):

Call the generated adjustSVGTagNameCase instead of building and looking
up a HashMap for the tag name case adjustment.

(WebCore::HTMLTreeBuilder::processStartTagForInBody):

Use TagName instead of comparing AtomStrings and use ElementName instead
of AtomString local names. Note the flow for applet/embed/object/marquee
is slightly changed due to the use of the switch statement now.

(WebCore::HTMLTreeBuilder::didCreateCustomOrFallbackElement):

Use element name argument, which is now unnecessary since HTMLStackItem
will extract the name itself.

(WebCore::HTMLTreeBuilder::processTemplateEndTag):
(WebCore::HTMLTreeBuilder::processEndOfFileForInTemplateContents):
(WebCore::HTMLTreeBuilder::processColgroupEndTagForInColumnGroup):
(WebCore::HTMLTreeBuilder::closeTheCell):
(WebCore::HTMLTreeBuilder::processStartTagForInTable):
(WebCore::HTMLTreeBuilder::processBodyEndTagForInBody):
(WebCore::HTMLTreeBuilder::callTheAdoptionAgency):
(WebCore::HTMLTreeBuilder::resetInsertionModeAppropriately):
(WebCore::HTMLTreeBuilder::processEndTagForInTableBody):
(WebCore::HTMLTreeBuilder::processEndTagForInRow):
(WebCore::HTMLTreeBuilder::processEndTagForInCell):
(WebCore::HTMLTreeBuilder::processEndTagForInBody):
(WebCore::HTMLTreeBuilder::processCaptionEndTagForInCaption):
(WebCore::HTMLTreeBuilder::processTrEndTagForInRow):
(WebCore::HTMLTreeBuilder::processTableEndTagForInTable):
(WebCore::HTMLTreeBuilder::processEndTagForInTable):
(WebCore::HTMLTreeBuilder::processEndTag):
(WebCore::HTMLTreeBuilder::insertPhoneNumberLink):
(WebCore::disallowTelephoneNumberParsing):
(WebCore::HTMLTreeBuilder::processCharacterBuffer):
(WebCore::HTMLTreeBuilder::processEndOfFile):
(WebCore::HTMLTreeBuilder::defaultForBeforeHTML):
(WebCore::HTMLTreeBuilder::defaultForBeforeHead):
(WebCore::HTMLTreeBuilder::defaultForInHead):
(WebCore::HTMLTreeBuilder::defaultForInHeadNoscript):
(WebCore::HTMLTreeBuilder::defaultForAfterHead):
(WebCore::HTMLTreeBuilder::processStartTagForInHead):
(WebCore::HTMLTreeBuilder::shouldProcessTokenInForeignContent):
(WebCore::HTMLTreeBuilder::processTokenInForeignContent):

Use TagNames and ElementNames instead of AtomStrings and QualifiedNames.

(WebCore::HTMLTreeBuilder::processStartTag):

Use TagNames and ElementNames instead of AtomStrings and QualifiedNames.
Call processFakeEndTag in a few places where we were constructing an
AtomHTMLToken locally and calling processEndTag.

(WebCore::HTMLTreeBuilder::processAnyOtherEndTagForInBody):

Split processAnyOtherEndTagForInBody into two cases: one for known
elements (common), and the other for unknown elements (rare) where we
have to delve into the Element on the HTMLStackItem to match its name.

* Source/WebCore/html/parser/TextDocumentParser.cpp:
(WebCore::TextDocumentParser::insertFakeFormattingElements):

Use TagName instead of AtomString for tag name.

* Source/WebCore/page/ProcessWarming.cpp:
(WebCore::ProcessWarming::initializeNames):

Initialize tag name strings.

Canonical link: https://commits.webkit.org/254007@main
